Sources/TuistDependencies/Carthage/CarthageInteractor.swift

```diff
@@ -0,0 +1,195 @@
+}
+
+// MARK: - Carthage Interactor
+
+public final class CarthageInteractor: CarthageInteracting {
+    private let fileHandler: FileHandling
+    private let carthageCommandGenerator: CarthageCommandGenerating
+    private let cartfileContentGenerator: CartfileContentGenerating
+
+    public init(
+        fileHandler: FileHandling = FileHandler.shared,
+        carthageCommandGenerator: CarthageCommandGenerating = CarthageCommandGenerator(),
+        cartfileContentGenerator: CartfileContentGenerating = CartfileContentGenerator()
+    ) {
+        self.fileHandler = fileHandler
+        self.carthageCommandGenerator = carthageCommandGenerator
+        self.cartfileContentGenerator = cartfileContentGenerator
+    }
+
+    public func install(dependenciesDirectory: AbsolutePath, method: InstallDependenciesMethod, dependencies: [CarthageDependency]) throws {
+        switch method {
+        case .fetch: logger.info("Start fetching Carthage dependencies.", metadata: .section)
+        case .update: logger.info("Start updating Carthage dependencies.", metadata: .section)
+        }
+
+        // check availability of `carthage`
+        guard canUseSystemCarthage() else {
+            throw CarthageInteractorError.carthageNotFound
+        }
+
+        // determine platforms
+        let platforms: Set<Platform> = dependencies
+            .reduce(Set<Platform>()) { platforms, dependency in platforms.union(dependency.platforms) }
+
+        try fileHandler.inTemporaryDirectory { temporaryDirectoryPath in
+            // prepare paths
+            let pathsProvider = PathsProvider(dependenciesDirectory: dependenciesDirectory, temporaryDirectoryPath: temporaryDirectoryPath)
+
+            // prepare for installation
+            try prepareForInstallation(pathsProvider: pathsProvider, dependencies: dependencies)
+
+            // create `carthage` shell command
+            let command = carthageCommandGenerator.command(method: method, path: temporaryDirectoryPath, platforms: platforms)
+
+            // run `carthage`
+            try System.shared.runAndPrint(command)
+
+            // post intallation actions
+            try postInstallationActions(pathsProvider: pathsProvider)
+        }
+
+        switch method {
+        case .fetch: logger.info("Carthage dependencies were fetched successfully.", metadata: .success)
+        case .update: logger.info("Carthage dependencies were updated successfully.", metadata: .success)
+        }
+    }
+
+    // MARK: - Installation
+
+    private func prepareForInstallation(pathsProvider: PathsProvider, dependencies: [CarthageDependency]) throws {
+        // copy build directory from previous run if exist
+        if fileHandler.exists(pathsProvider.destinationCarthageDirectory) {
+            try copyDirectory(from: pathsProvider.destinationCarthageDirectory, to: pathsProvider.temporaryCarthageBuildDirectory)
```

Comment:
Why are we doing this? How does Carthage use any existing frameworks?

Reply:
I need it to prevent frameworks recompiling.